jupyterhub/app.py

```diff
@@ -1145,7 +1145,8 @@ def init_services(self):
             if service.managed:
                 if not service.api_token:
                     # generate new token
-                    service.api_token = service.orm.new_api_token()
+                    # TODO: revoke old tokens?
```

Comment:
Are there use cases where a managed service would need to keep an old api_token?

Reply:
Not one generated here. I think the main bit missing is that we don't track whether this happened, so we don't have enough info to clean it up afterward.